fuzzer/src/main.rs

```diff
@@ -0,0 +1,443 @@
+//! The first pass doesn't perform any outlier detection / handling.
+//! This is because outlier handling is pretty slow.
+//! The first pass just assumes there aren't any outliers, calculates `SAMPLE_SIZE` points and
+//! applies the scoring functions to them.
+//! This filters out most patterns already.
+//! If however possible non-linear behaviour is found, we perform the second pass.
+//! In the second pass, we double the `SAMPLE_SIZE`, because more points make for higher precision
+//! in the scoring functions.
+//! Additionally, we assume that the (len, time)-pairs interpreted as coordinates on a graph  are
+//! strictly monotonously increasing.
+//! If we encounter a time-value, which is larger than the next time value, that'll probably be an
+//! outlier.
+//! As such, if we encounter such a value, we'll re-time it.
+//! Unfortunately, again, we have an edge-case.
+//! When testing substrings of the input, we might not always end up after the whole pattern,
+//! but instead in the middle of it.
+//! Sometimes pulldown-cmark parses things differently if the strings ends differently.
+//! For one example see <https://github.com/raphlinus/pulldown-cmark/issues/287>.
+//! Thus we also need to detect these case.
+//! Therefore we only remove a maximum of 10 outliers.
+//! If more outliers are detected, measurements are aborted and an according message is printed.
+//!
+//! Additionally, we abort if parsing took longer than `MAX_MILLIS`.
+//! If we tested a substring of the whole input, and already exceeded `MAX_MILLIS`, we shouldn't
+//! continue testing with longer inputs, as we might have super-linear behaviour.
+//! Instead, we again print an according message and abort.
+
+use std::time::{Duration, Instant};
+use std::panic;
+use std::sync::{Mutex, atomic::{AtomicU64, Ordering}};
+use std::env;
+use std::io::{self, BufRead};
+use std::iter;
+use std::process::Command;
+use std::os::unix::process::CommandExt;
+
+use itertools::Itertools;
+use pulldown_cmark::{Parser, Options};
+use rand::{SeedableRng, distributions::{Distribution, Uniform}};
+use rand_xoshiro::Xoshiro256Plus;
+use crossbeam_utils::thread;
+
+mod literals;
+mod black_box;
+mod scoring;
+mod clock;
+
+use clock::{Clock, ThreadCpuTime};
+
+/// After how many patterns each threads prints status reports including throughput.
+const BATCH_SIZE: usize = 10_000;
+/// Length of number of literals per combination.
+///
+/// For example given the literals `a` and `b`, with a value of 4, `aaab` is a possible pattern.
+const COMBINATIONS: usize = 6;
+/// If parsing takes longer than this many milliseconds, it is assumed to be non-linear and
+/// parsing is aborted.
+const MAX_MILLIS: u128 = 500;
+/// Byte-length of maximum pattern repetitions used as input to pulldown-cmark.
+const NUM_BYTES: usize = 256*1024;
+/// Number of samples per pattern tested. Higher number increases precision, but reduces the throughput.
+const SAMPLE_SIZE: usize = 5;
```

Comment:
The higher the sample size, the longer the analysis takes. It takes O(SAMPLE_SIZE²) due to the even distribution of measurement input regarding the input length. 5 for the first pass and 2*5 for the second pass seemed ok, especially because we verify all outputs multiple times afterwards with the `./retest-output` script. Obviously better tweaking of those numbers will probably achieve better results and a lower false-positive rate, but the current numbers seem good enough for my machine.

There are probably lots of easy ways to improve the mechanics used for superlinear growth detection. But I didn't want to write a paper about this (yet) ;)